grr/server/grr_response_server/bin/grrafana.py

```diff
@@ -0,0 +1,165 @@
+    return JSONResponse(response=json.dumps(response), mimetype="application/json")
+
+  def onQuery(self, request: JSONRequest) -> JSONResponse:
+    """Given a client ID as a Grafana variable and targets (resource usages),
+    returns datapoints in a format Grafana can interpret."""
+    request_json_format = request.json
+    requested_client_id = extractClientIdFromVariable(request_json_format)  # There must be a ClientID variable declated in Grafana.
+    requested_targets = [entry["target"] for entry in request_json_format["targets"]]
+    response = fetchDatapointsForTargets(requested_client_id, request_json_format["maxDataPoints"], requested_targets)
+    return JSONResponse(response=json.dumps(response), mimetype="application/json")
+
+  def onAnnotations(self, request: JSONRequest) -> JSONResponse:
+    pass
+
+
+def fetchClientIds() -> List[Text]:
+  """Fetches GRR client IDs that have resource usage records in Fleetspeak database."""
+  return fleetspeak_utils.GetClientIdsFromFleetspeak()
+
+
+Datapoint = Tuple[float, int]
+Datapoints = List[Datapoint]
+TargetWithDatapoints = Dict[Text, Datapoints]
+def fetchDatapointsForTargets(client_id: Text, limit: int, targets: List[Text]) -> List[TargetWithDatapoints]:
+  """Fetches a list of <datapoint, timestamp> tuples for each target metric from Fleetspeak database."""
+  records_list = fleetspeak_utils.FetchClientResourceUsageRecordsFromFleetspeak(client_id, limit)
+  response = list()
+  for target in targets:
+    datapoints_for_single_target = createDatapointsForTarget(target, records_list)
+    target_datapoints_dict = typing.cast(TargetWithDatapoints, {"target": target, "datapoints": datapoints_for_single_target})
+    response.append(target_datapoints_dict)
+  return response
+
+
+def createDatapointsForTarget(target: Text, records_list: List[ClientResourceUsageRecord]) -> Datapoints:
+  if target == "mean_user_cpu_rate":
+    record_values = [record.mean_user_cpu_rate for record in records_list]
+  elif target == "max_user_cpu_rate":
+    record_values = [record.max_user_cpu_rate for record in records_list]
+  elif target == "mean_system_cpu_rate":
+    record_values = [record.mean_system_cpu_rate for record in records_list]
+  elif target == "max_system_cpu_rate":
+    record_values = [record.max_system_cpu_rate for record in records_list]
+  elif target == "mean_resident_memory_mib":
+    record_values = [record.mean_resident_memory_mib for record in records_list]
+  elif target == "max_resident_memory_mib":
+    record_values = [record.max_resident_memory_mib for record in records_list]
+  return [(v, r.server_timestamp.seconds * 1000) for (v, r) in zip(record_values, records_list)]
+
+
+def extractClientIdFromVariable(req: JSONRequest) -> Text:
+  """Extracts the client ID from a Grafana JSON request."""
+  # Based on an assumption that there is only one Grafana variable.
+  return next(iter(req["scopedVars"].values()))["value"]
```

Comment:
This is very complicated. Please rewrite as:
```
scoped_vars_values = list(req["scopedVars"].values())
return scoped_vars_values[0]["value"]
```

Quick question - I'd expect the client id variable to have a name. Instead of using `.values` can we just access the dictionary by key? I.e. `req["scopedVars"]["clientId"]["value"]`?

Reply:
That's a small issue. You're right that the Client ID variable has a name, the problem is it is up to the user to choose its name on creation within the Grafana UI. You can check out the contents of a sample `req["scopedVars"]` dictionary here:
```
{
'myname': {'text': 'C.<some_client_id>', 'value': 'C.<some_client_id>'}, 
'__interval': {'text': '30m', 'value': '30m'}, 
'__interval_ms': {'text': '1800000', 'value': 1800000}
}
```
We can ask users to name the variable with a specific name with documentation, but I did not think it was optimal. Do you have any ideas?

Reply:
Following an offline discussion: we will be the ones providing the Grafana dashboard to display clients load, so I'd expect that we would be able to pre-configure the variable name in the dashboard configuration. If this is correct, then there's no issue and we can rely on the variable name being predefined.

But in any case, even if the user has to define a variable by hand, it's much cleaner to document the intended name and ask the user to use it. The current code would effectively break if there's more than one variable defined, which feels very fragile.

Reply:
As you suspected, we will indeed be able to pre-configure the variable name in the Grafana dashboard when we supply it, within the [textbox variable](https://grafana.com/docs/grafana/latest/variables/variable-types/add-text-box-variable/#add-a-text-box-variable) as discussed in the [other conversation](https://github.com/google/grr/pull/832#discussion_r478403721). Therefore, the issue with the code here is resolved.

However, there is one minor thing left to discuss. The library [grafanalib](https://github.com/weaveworks/grafanalib), which we also use within GRR in the [`grr_grafanalib_dashboards`](https://github.com/google/grr/tree/master/monitoring/grafana) package, [does support adding variables](https://grafanalib.readthedocs.io/en/latest/api/grafanalib.html#grafanalib.core.Template) to scripted dashboards, but does not support the specific type of variable we need (textbox).
I don't suspect it to be a problem when we get to the dashboards in a later PR, but it's something to be aware off, since we will essentially be forced to work around the API of grafanalib (we did in it [other cases](https://github.com/google/grr/blob/b472626b6ff4d05b24269a7ce52419d00a847607/monitoring/grafana/grr_grafanalib_dashboards/dashboards_to_generate/fleetspeak.dashboard.py#L52) before, though not as significant as this one).

Reply:
Thanks for researching this. I'd look into adding this functionality to grafanalib and submitting it upstream. If adding such support is not too much work this would be the best option. Otherwise, yes, we can work around the API of grafanlib if needed.